source/common/http/conn_manager_impl.cc

```diff
@@ -793,17 +796,23 @@ void ConnectionManagerImpl::ActiveStream::decodeData(ActiveStreamDecoderFilter*
     }
   }
 
+  // If trailers were adding during decodeData we need to trigger decodeTrailers in order
+  // to allow filters to process the trailers.
   if (end_stream && request_trailers_) {
     decodeTrailers(filter, *request_trailers_);
   }
 }
 
-void ConnectionManagerImpl::ActiveStream::addDecodedTrailers(HeaderMapPtr&& trailers) {
+HeaderMap& ConnectionManagerImpl::ActiveStream::addDecodedTrailers() {
   if (state_.filter_call_state_ & FilterCallState::LastDataFrame) {
-    request_trailers_ = std::move(trailers);
+    if (request_trailers_) {
+      throw std::logic_error("decodedTrailers added more than once");
```

Comment:
So what will catch these exceptions?  Perhaps they should be CodecProtocolException so that the existing codec code will catch them and terminate the connection?

Reply:
I picked a non-EnvoyException exception as per @mattklein123 suggestion in a previous comment (https://github.com/envoyproxy/envoy/pull/3980#discussion_r207596132). If we want this to fail more gracefully then I'm happy to do so, just want to make sure we're all in agreement before switching over.

Reply:
IMO this is the right exception type so that we don't catch it and crash/core dump. I think a logic error like this should be caught during development time and should be very obvious?

Reply:
Sorry, to be clear,  you think this exception should not be caught?   I grant with our current example filters it should be fairly obvious, but once folks have custom filters which do non-trivial async work which may have timing invariants, and/or we make the pipeline more complicated with internal redirects, I think it gets less obvious.  I'd strongly prefer developer errors cause disconnects and not crashes.

Reply:
Yes, I think it should not be caught. If you catch this exception you will need to add stats and logging to make it obvious what is happening otherwise it will be incredibly confusing. Basically, it's a bunch of error handling and operational work that IMO should not happen in practice. If there is disagreement on using an exception for this I would switch to RELEASE_ASSERT and make it crash.

Reply:
Can we compromise on ASSERT for documenting this is an invariant since this isn't likely going to be due to memory corruption?

Reply:
Sure that's fine with me for now, though, I do think we need to help filter writers catch programming errors that might not be caught in a release build. But, there are already TODOs about this (and other ASSERTs) so we can cover this in a future change for sure.

Reply:
So the tl;dr here is that I should be using ASSERTs instead of exceptions, deferring better error handling to a future change?

@alyssawilk Given this, would you still want integration tests? It'd only test the happy path (unless we can test asserts somehow?), but it might still be worth having?

Reply:
Yeah, I think given this we're just testing the happy path.  One theoretically could do a debug death test for broken filters but even I think that'd be overkill :-)

Reply:
We've got a bunch of tests which add filters (see config_helper_.addFilter()) so what we'll need to do is add a test (probably in http_integration.cc and then called from the various http[2][_upstream]_integration_test.cc) which does makeRequestWithBody() and then verifies header/body/data get proxied downstream and then the flip side, upstream_request_->encodeHeaders() and encodeData with trailers being received by H2 downstream.  

The trickier bit is we don't have a sample filter which does this, so you'll need to add a test filter which always adds trailers.   In the long run I'd like a test filter factory which creates MockStreamFilter which we can use that for all sorts of custom test behavior but I think that's out of scope for this PR :-)  If you do the simple one I'll hopefully get some free time to convert it to the fancy version later in.

Actually it occurs to me thinking about makeRequestWithBody vs makeHeaderOnlyRequest.... we allow adding trailers on decodeData(end_stream = true) but should we also allow it in decodeHeaders (end_stream=true)?    It seems more consistent for H2 but if we're really only wanting this for gRPC *and* we think we won't hit gRPC with no body where we want to add trailers, I'd be OK documenting it as a known limitation someone else can fix when/if they need it.

Reply:
Thanks for the pointers, I'll try to get some tests set up.

I can't really think of a use case of adding trailers to request/response with no DATA frames (how is that even different from just adding additional headers?). I'd prefer to to just leave it as it is right now. Adding it in later should be fairly straightforward once all the unit tests/integration tests are in place.